File: dasherdancer/src/main/java/com/google/android/apps/santatracker/dasherdancer/CharacterAdapter.java
/*
 * Copyright (C) 2016 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.santatracker.dasherdancer;


import android.support.v4.view.PagerAdapter;
import android.util.Log;
import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewGroup.LayoutParams;
import android.widget.ImageView;

import java.lang.*;

public class CharacterAdapter extends PagerAdapter {

	private static final String LOG_TAG = CharacterAdapter.class.getSimpleName();
	
	private static final int[] mBackgrounds = new int[]{
		R.color.bg_blue,R.color.bg_pink,R.color.bg_purple,
		R.color.bg_green,R.color.bg_red,R.color.bg_orange,
		R.color.bg_yellow
	};
	
	private SparseArray<ImageView> mViews;
	private Character[] mCharacters;
	
	public CharacterAdapter(Character[] characters) {
		mCharacters = characters;
		mViews = new SparseArray<ImageView>();
	}
	
	@Override
	public void destroyItem(ViewGroup container, int position, Object object) {
		if (position < mCharacters.length) {
			View view = mViews.get(position);
			if ((view == null) && (object != null)) {
				try {
					view = (ImageView)object;
				} catch (ClassCastException e) {
					Log.e(LOG_TAG, "Wrong type of object supplied to destroyItem.");
				}
			}
			
			if (view != null) {
				container.removeView(view);
			}
			
			mViews.remove(position);
		}
	}
	
	@Override
	public int getCount() {
		if(mCharacters != null) {
			return mCharacters.length;
		}
		return 0;
	}

	@Override
	public Object instantiateItem(ViewGroup container, int position) {
		ImageView view = null;
		
		if (position < mCharacters.length) {
			if (mViews.get(position) != null) {
				view = mViews.get(position);
			} else {
				view = new FrameAnimationView(container.getContext());
				view.setScaleType(ImageView.ScaleType.MATRIX);
				//Load the first idle frame.
				view.setBackgroundResource(mBackgrounds[position]);
				view.setTag(position);
				mViews.put(position, view);
			}
			
			LayoutParams lp = new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT);
			container.addView(view, lp);
		}
		
		return view;
	}
	
	@Override
	public boolean isViewFromObject(View arg0, Object arg1) {
		return arg0 == arg1;
	}

}

File: dasherdancer/src/main/java/com/google/android/apps/santatracker/dasherdancer/CharacterAdapter.kt
/*
 * Copyright 2019. Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.santatracker.dasherdancer

import android.util.SparseArray
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams
import android.widget.ImageView
import androidx.viewpager.widget.PagerAdapter
import com.google.android.apps.santatracker.util.SantaLog

class CharacterAdapter(private val characters: Array<Character>) : PagerAdapter() {

    private val views: SparseArray<ImageView> = SparseArray()

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        if (position < characters.size) {
            var view: View? = views.get(position)
            if (view == null) {
                try {
                    view = `object` as ImageView
                } catch (e: ClassCastException) {
                    SantaLog.e(LOG_TAG, "Wrong type of object supplied to destroyItem.")
                }
            }

            if (view != null) {
                container.removeView(view)
            }

            views.remove(position)
        }
    }

    override fun getCount(): Int {
        return characters.size
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        var view: ImageView? = null

        if (position < characters.size) {
            if (views.get(position) != null) {
                view = views.get(position)
            } else {
                view = FrameAnimationView(container.context)
                view.scaleType = ImageView.ScaleType.MATRIX
                // Load the first idle frame.
                view.setBackgroundResource(mBackgrounds[position])
                view.tag = position
                views.put(position, view)
            }

            val lp = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
            container.addView(view, lp)
        }

        return view!!
    }

    override fun isViewFromObject(arg0: View, arg1: Any): Boolean {
        return arg0 === arg1
    }

    companion object {

        private val LOG_TAG = CharacterAdapter::class.java.simpleName

        private val mBackgrounds = intArrayOf(R.color.bg_blue, R.color.bg_pink, R.color.bg_purple, R.color.bg_green, R.color.bg_red, R.color.bg_orange, R.color.bg_yellow)
    }
}
